### PR TITLE
Eigen decomp material

### LINKED ISSUE #29179: eigen decomposition of rank two material property
### Motivation

I want to output the eigen decomposition of a rank two material property.  I want this to be saved to a material property as a vector and value for each eigenvector.

### Design

I will make it a material that consumes another material that is a rank two tensor.

### Impact

This will provide a nice visualization for rank two tensors that will help identify the directions and magnitude of rank two tensors in solid_mechanics problems like stress and strain.

### PR DESCRIPTION
**Pull request by lynnmunday**
closes #29179

### CODE REVIEW

**Comment by GiudGiud** [comment]
@lynnmunday who's to be the reviewer here? To assign it

**Comment by lynnmunday** [comment]
@dschwen will you review?   or @maxnezdyur

File: modules/solid_mechanics/doc/content/source/materials/EigenDecompMaterial.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# EigenDecompMaterial
```
Comment (dschwen) [inline modules/solid_mechanics/doc/content/source/materials/EigenDecompMaterial.md:1]: Please rename this to `EigenDecompositionMaterial`

File: modules/solid_mechanics/doc/content/source/materials/EigenDecompMaterial.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+!syntax description /Materials/EigenDecompMaterial
+
+## Description
+
+This class reads in a Rank two tensor material given by [!param](/Materials/EigenDecompMaterial/rank_two_tensor) and performs an eigendecomposition on it.  The results of the decomposition are scalar materials named `max_eigen_value`, `mid_eigen_value`, and `min_eigen_value` and the corresponding vector material properties named `max_eigen_vector`, `mid_eigen_vector`, and `min_eigen_vector`. These names can be preceeded by the [!param](/Materials/EigenDecompMaterial/base_name) to allow for more than one `EigenDecompMaterial` per block.  These material properties can be output using the Material Outputs system with the [!param](/Materials/EigenDecompMaterial/output_properties) as shown in the below example.  This material is useful for visualizing the elemental maximum principal stress and direction as a vector.
```
Comment (dschwen) [inline modules/solid_mechanics/doc/content/source/materials/EigenDecompMaterial.md:7]:
```suggestion
This class reads in a `RankTwoTensor` material given by [!param](/Materials/EigenDecompMaterial/rank_two_tensor) and performs an eigendecomposition on it.  The results of the decomposition are scalar materials named `max_eigen_value`, `mid_eigen_value`, and `min_eigen_value` and the corresponding vector material properties named `max_eigen_vector`, `mid_eigen_vector`, and `min_eigen_vector`. These names can be preceeded by the [!param](/Materials/EigenDecompMaterial/base_name) to allow for more than one `EigenDecompMaterial` per block.  These material properties can be output using the Material Outputs system with the [!param](/Materials/EigenDecompMaterial/output_properties) as shown in the below example.  This material is useful for visualizing the elemental maximum principal stress and direction as a vector.
```

File: modules/solid_mechanics/include/materials/EigenDecompMaterial.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Material.h"
+#include "RankTwoTensorForward.h"
+/**
+ * Perform eigen decomposition on Rank two tensor
+ */
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/EigenDecompMaterial.h:16]:
```suggestion

/**
 * Perform eigendecomposition on a RankTwoTensor material property
 */
```

File: modules/solid_mechanics/include/materials/EigenDecompMaterial.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+class EigenDecompMaterialTempl : public Material
+{
+public:
+  static InputParameters validParams();
+
+  EigenDecompMaterialTempl(const InputParameters & parameters);
+
+  virtual void computeQpProperties() override;
+
+protected:
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/EigenDecompMaterial.h:27]:
```suggestion
protected:
    virtual void computeQpProperties() override;
```

this method is protected in the base class and should stay that way

File: modules/solid_mechanics/src/materials/EigenDecompMaterial.C [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+registerMooseObject("SolidMechanicsApp", EigenDecompMaterial);
+registerMooseObject("SolidMechanicsApp", ADEigenDecompMaterial);
+
+template <bool is_ad>
+InputParameters
+EigenDecompMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription(
+      "Creates material with eigenvalues and eigenvectors of a symmetric rank two tensor.");
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/EigenDecompMaterial.C:22]:
```suggestion
      "Emits material properties for the eigenvalues and eigenvectors of a symmetric rank two tensor.");
```

The fact that it has to be symmetric should be mentioned in the documentation. Also what happens if the tensor is not symmetric? You could just return the real part of the EVs then...

File: modules/solid_mechanics/src/materials/EigenDecompMaterial.C [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    _tensor(getGenericMaterialProperty<RankTwoTensor, is_ad>("rank_two_tensor")),
+    _max_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "max_eigen_vector")),
+    _mid_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "mid_eigen_vector")),
+    _min_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "min_eigen_vector")),
+    _max_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "max_eigen_value")),
+    _mid_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "mid_eigen_value")),
+    _min_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "min_eigen_value"))
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/EigenDecompMaterial.C:45]: This naming scheme assumes that `LIBMESH_DIM` is always 3. Realistically that's probably required in MOOSE, but yet we do have over 1600 checks for `LIBMESH_DIM` in our code. I suggest checking the dimension.

File: modules/solid_mechanics/test/tests/eigen_decomp_material/pipe_2d.i [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    [QuasiStatic]
+      [all]
+        strain = SMALL
+        add_variables = true
+        new_system = true
+        formulation = TOTAL
+        volumetric_locking_correction = false
+        # generate_output = "cauchy_stress_xx cauchy_stress_yy cauchy_stress_zz cauchy_stress_xy "
+        #                   "cauchy_stress_xz cauchy_stress_yz strain_xx strain_yy strain_zz strain_xy "
+        #                   "strain_xz strain_yz vonmises_cauchy_stress"
```
Comment (dschwen) [inline modules/solid_mechanics/test/tests/eigen_decomp_material/pipe_2d.i:41]:
```suggestion
```

File: modules/solid_mechanics/test/tests/eigen_decomp_material/prescribed_strain_3D.i [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    variable = disp_x
+    use_xyzt = true
+    expression = "if(t<4.1,4e-1*x*t,x)"
+  []
+  [disp_y]
+    execute_on = 'TIMESTEP_BEGIN'
+    type = ParsedAux
+    variable = disp_y
+    use_xyzt = true
+    expression = "if(t<4.1,3e-1*y*t^2,1e-1*y*t+1e-1*x*t)"
```
Comment (dschwen) [inline modules/solid_mechanics/test/tests/eigen_decomp_material/prescribed_strain_3D.i:40]: I'd add a comment to all those expressions explaining what the intent is

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/solid_mechanics/src/materials/EigenDecompMaterial.C (hunk 5 above)
Comment (lynnmunday) [reply modules/solid_mechanics/src/materials/EigenDecompMaterial.C:22]: It gives the wrong answer so I added an error check.

**Review by dschwen — APPROVED** [review]

File: modules/solid_mechanics/test/tests/eigen_decomp_material/tests [hunk 9]
```diff
@@ -1,13 +1,12 @@
 [Tests]
   design = 'source/materials/EigenDecompositionMaterial.md'
   issues = '#29179'
-  requirement = 'The system shall output an eigen decomposition of a symmetric rank two tensor '
```
Comment (dschwen) [inline modules/solid_mechanics/test/tests/eigen_decomp_material/tests]: This would have worked if you had added one more level to set up a group of tests.

Comment (lynnmunday) [reply modules/solid_mechanics/test/tests/eigen_decomp_material/tests]: I was going to but I didn't know what to name it...... [Tests][tests]

Comment (GiudGiud) [reply modules/solid_mechanics/test/tests/eigen_decomp_material/tests]: [Tests]
  design = 'source/materials/EigenDecompositionMaterial.md'
  issues = '#29179'
  [eigen_decomp_output]
     [this case]
      ...
      [this other one]

    []
  []
```

**Pull request merged**